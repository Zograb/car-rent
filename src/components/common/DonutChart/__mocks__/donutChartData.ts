const donutChartData = [
  {
    name: 'Part 1',
    value: 360,
    color: '#52C93F',
  },
  {
    name: 'Part 2',
    value: 230,
    color: '#006AFF',
  },
  {
    name: 'Part 3',
    value: 150,
    color: '#FF2727',
  },
  {
    name: 'Part 4',
    value: 230,
    color: '#FF2727',
  },
  {
    name: 'Part 5',
    value: 150,
    color: '#FFFF00',
  },
];

export default donutChartData;
