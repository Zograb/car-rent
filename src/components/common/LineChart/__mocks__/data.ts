export const lineChartData = [
  { label: 'Label 1', value1: 222, value2: 450 },
  { label: 'Label 2', value1: 120, value2: 250 },
  { label: 'Label 3', value1: 340, value2: 320 },
  { label: 'Label 4', value1: 190, value2: 400 },
  { label: 'Label 5', value1: 520, value2: 150 },
  { label: 'Label 6', value1: 320, value2: 370 },
  { label: 'Label 7', value1: 300, value2: 220 },
  { label: 'Label 8', value1: 600, value2: 200 },
  { label: 'Label 9', value1: 250, value2: 450 },
  { label: 'Label 10', value1: 450, value2: 250 },
];
