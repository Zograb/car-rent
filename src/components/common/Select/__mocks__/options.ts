const options = [
  { value: '6465', label: '6465' },
  { value: '5665', label: '5665' },
  { value: '1755', label: '1755' },
  { value: '4532', label: '4532' },
  { value: '5543', label: '5543' },
];

export default options;
