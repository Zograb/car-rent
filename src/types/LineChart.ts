export interface LineChartLine {
  key: string;
  drawAreaGradient?: boolean;
  color?: string;
  dashed?: boolean;
}
