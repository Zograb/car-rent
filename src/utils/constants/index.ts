export * from './enums';
export * from './tailwindConfig';
