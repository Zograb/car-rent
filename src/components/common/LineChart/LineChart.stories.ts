import type { Meta, StoryObj } from '@storybook/react';

import { LineChart } from './LineChart';
import { lineChartData } from './__mocks__/data';
import { COLORS } from '@/constants/tailwindConfig';

const meta: Meta<typeof LineChart> = {
  title: 'Line Chart',
  component: LineChart,
  tags: ['autodocs'],
};

export default meta;
type Story = StoryObj<typeof LineChart>;

export const Basic: Story = {
  args: {
    data: lineChartData,
    labelsDataKey: 'label',
    lines: [{ key: 'value1' }, { key: 'value2' }],
  },
};

export const WithFixedColors: Story = {
  args: {
    data: lineChartData,
    labelsDataKey: 'label',
    lines: [
      {
        key: 'value1',
        color: COLORS['primary'],
      },
      {
        key: 'value2',
        color: COLORS['black'],
      },
    ],
  },
}

export const WithAreaGradient: Story = {
  args: {
    data: lineChartData,
    labelsDataKey: 'label',
    lines: [
      {
        key: 'value1',
        drawAreaGradient: true,
      },
      {
        key: 'value2',
        drawAreaGradient: true,
      },
    ],
  },
}

export const WithDashedLine: Story = {
  args: {
    data: lineChartData,
    labelsDataKey: 'label',
    lines: [
      {
        key: 'value1',
        drawAreaGradient: true,
      },
      {
        key: 'value2',
        dashed: true,
      },
    ],
  },
}
