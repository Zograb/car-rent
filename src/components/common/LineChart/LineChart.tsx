import type { CategoricalChartProps } from 'recharts/types/chart/generateCategoricalChart';
import type { LineChartLine } from '@/types';
import { Fragment, memo } from 'react';
import {
  type YAxisProps,
  type XAxisProps,
  type TooltipProps,
  Area,
  ComposedChart,
  Line,
  XAxis,
  YAxis,
  Tooltip,
  ResponsiveContainer,
} from 'recharts';
import 'twin.macro';

// Utils
import { getRandomColor } from '@/utils';

export interface LineChartProps extends CategoricalChartProps {
  // key to determine which field is label
  labelsDataKey: string;

  // array of config objects to render lines
  lines: LineChartLine[];

  // chart data
  // eslint-disable-next-line
  data: any[];

  // config for chart Y axis
  yAxisProps?: YAxisProps;

  // config for chat X axis
  xAxisProps?: XAxisProps;

  // config for tooltip
  tooltipProps?: TooltipProps<number, string>;
}

export const LineChart = memo(({
  data,
  labelsDataKey,
  lines,
  yAxisProps,
  xAxisProps,
  tooltipProps,
}: LineChartProps) => (
  <div tw="overflow-x-scroll sm:overflow-x-auto">
    <ResponsiveContainer width="99%" height={250} minWidth={300}>
      <ComposedChart data={data} margin={{ bottom: 25, top: 10 }}>
        <XAxis
          dataKey={labelsDataKey}
          axisLine={false}
          tickLine={false}
          dy={25}
          {...xAxisProps}
        />
        <YAxis
          axisLine={false}
          tickLine={false}
          dx={-10}
          {...yAxisProps}
        />
        <Tooltip {...tooltipProps} />
        {lines.map((line) => {
          const color = line.color || getRandomColor();
          return (
            <Fragment key={line.key}>
              <Line
                type="monotone"
                dataKey={line.key}
                stroke={color}
                strokeDasharray={line.dashed ? '8 8' : ''}
              />
              {line.drawAreaGradient && (
                <Fragment>
                  <Area
                    type="monotone"
                    dataKey={line.key}
                    stroke={color}
                    strokeWidth={2}
                    fillOpacity={1}
                    fill={`url(#color-${line.key})`}
                  />
                  <defs>
                    <linearGradient
                      id={`color-${line.key}`}
                      x1="0"
                      y1="0"
                      x2="0"
                      y2="1"
                    >
                      <stop offset="5%" stopColor={color} stopOpacity={0.3}/>
                      <stop offset="95%" stopColor={color} stopOpacity={0}/>
                    </linearGradient>
                  </defs>
                </Fragment>
              )}
            </Fragment>
          );
        })}
      </ComposedChart>
    </ResponsiveContainer>
  </div>
));

LineChart.displayName = 'LineChart ';
