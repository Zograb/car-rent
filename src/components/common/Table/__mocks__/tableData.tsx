import { FiTrash2 } from 'react-icons/fi';
import tw from 'twin.macro';

// Components
import { Button } from '@/components';

// Utils
import { formatNumberToCurrency } from '@/utils';

const row = {
  number: <div className="text-sm">01</div>,
  id: '12345',
  user: 'John Doe',
  status: 'Approved',
  earning: formatNumberToCurrency(3250),
  actions: (
    <div className="flex justify-end items-center">
      <Button size="small" className="mr-2 text-xl" icon={<FiTrash2 />} iconButton />
      <Button size="small">Details</Button>
    </div>
  ),
}

export const rows = [row, row, row, row];
export const columns = [
  { key: 'id', label: 'ID' },
  { key: 'user', label: 'User' },
  { key: 'status', label: 'Status' },
  { key: 'earning', label: 'Earning' },
  { key: 'actions', label: '' },
];
