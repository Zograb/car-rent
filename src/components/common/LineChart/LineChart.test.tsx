import { cloneElement, type ReactElement } from 'react';
import { render } from '@testing-library/react';
import recharts from 'recharts';

// Components
import LineChart from './LineChart';

// Mocks
import { lineChartData } from './__mocks__/data';
import {
  mockResizeObserverBefore,
  mockResizeObserverAfter,
} from '@/__mocks__';

// Constants
import { COLORS } from '@/constants/tailwindConfig';

const ResponsiveContainer = ({ children, height }: { children: ReactElement, height: number }) =>
  cloneElement(children, { width: 500, height })

// Remove responsive container and render line chart directly
// Otherwise in test envirinment it will not render chart at all
jest.mock('recharts', () => {
  const OriginalModule: typeof recharts = jest.requireActual('recharts');

  return {
    ...OriginalModule,
    ResponsiveContainer,
  };
});

describe('Line Chart', () => {
  beforeEach(mockResizeObserverBefore);

  afterEach(mockResizeObserverAfter);

  it('should match snapshot', () => {
    const { container } = render(
      <LineChart
        data={lineChartData}
        labelsDataKey="date"
        lines={[
          { key: 'value', drawAreaGradient: true, color: COLORS['primary'] },
          {
            key: 'previousYearValue',
            drawAreaGradient: false,
            dashed: true,
            color: COLORS['gray3'],
          },
        ]}
      />
    );
    expect(container).toMatchSnapshot();
  });

  it('should render line chart', () => {
    const { getByRole } = render(
      <LineChart
        data={lineChartData}
        labelsDataKey="date"
        lines={[
          {
            key: 'value',
            drawAreaGradient: true,
            color: COLORS['primary'],
          },
          {
            key: 'previousYearValue',
            drawAreaGradient: false,
            dashed: true,
            color: COLORS['gray3'],
          },
        ]}
      />
    );
    const lineChart = getByRole('region');
    expect(lineChart).toBeInTheDocument();
  });

  it('should render line chart with custom ticks', () => {
    const { container } = render(
      <LineChart
        data={lineChartData}
        labelsDataKey="date"
        lines={[
          {
            key: 'value',
            drawAreaGradient: true,
            color: COLORS['primary'],
          },
          {
            key: 'previousYearValue',
            drawAreaGradient: false,
            dashed: true,
            color: COLORS['gray3'],
          },
        ]}
        xAxisProps={{
          tickFormatter: () => 'xAxisTick',
        }}
      />
    );
    const xAxisTick = container.querySelector('.recharts-cartesian-axis-tick-value');
    expect(xAxisTick).toHaveTextContent('xAxisTick');
  })
});
