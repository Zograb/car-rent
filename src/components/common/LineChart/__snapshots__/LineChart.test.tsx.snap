// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Line Chart should match snapshot 1`] = `
.c0 {
  overflow-x: scroll;
}

@media (min-width: 640px) {
  .c0 {
    overflow-x: auto;
  }
}

<div>
  <div
    class="c0"
  >
    <div
      class="recharts-wrapper"
      role="region"
      style="position: relative; cursor: default; width: 500px; height: 250px;"
    >
      <svg
        class="recharts-surface"
        height="250"
        viewBox="0 0 500 250"
        width="500"
      >
        <title />
        <desc />
        <defs>
          <clippath
            id="recharts1-clip"
          >
            <rect
              height="185"
              width="440"
              x="60"
              y="10"
            />
          </clippath>
        </defs>
        <g
          class="recharts-layer recharts-cartesian-axis recharts-xAxis xAxis"
        >
          <g
            class="recharts-cartesian-axis-ticks"
          >
            <g
              class="recharts-layer recharts-cartesian-axis-tick"
            >
              <text
                class="recharts-text recharts-cartesian-axis-tick-value"
                fill="#666"
                height="30"
                orientation="bottom"
                stroke="none"
                text-anchor="middle"
                width="440"
                x="60"
                y="228"
              >
                <tspan
                  dy="0.71em"
                  x="60"
                />
              </text>
            </g>
            <g
              class="recharts-layer recharts-cartesian-axis-tick"
            >
              <text
                class="recharts-text recharts-cartesian-axis-tick-value"
                fill="#666"
                height="30"
                orientation="bottom"
                stroke="none"
                text-anchor="middle"
                width="440"
                x="108.88888888888889"
                y="228"
              >
                <tspan
                  dy="0.71em"
                  x="108.88888888888889"
                />
              </text>
            </g>
            <g
              class="recharts-layer recharts-cartesian-axis-tick"
            >
              <text
                class="recharts-text recharts-cartesian-axis-tick-value"
                fill="#666"
                height="30"
                orientation="bottom"
                stroke="none"
                text-anchor="middle"
                width="440"
                x="157.77777777777777"
                y="228"
              >
                <tspan
                  dy="0.71em"
                  x="157.77777777777777"
                />
              </text>
            </g>
            <g
              class="recharts-layer recharts-cartesian-axis-tick"
            >
              <text
                class="recharts-text recharts-cartesian-axis-tick-value"
                fill="#666"
                height="30"
                orientation="bottom"
                stroke="none"
                text-anchor="middle"
                width="440"
                x="206.66666666666666"
                y="228"
              >
                <tspan
                  dy="0.71em"
                  x="206.66666666666666"
                />
              </text>
            </g>
            <g
              class="recharts-layer recharts-cartesian-axis-tick"
            >
              <text
                class="recharts-text recharts-cartesian-axis-tick-value"
                fill="#666"
                height="30"
                orientation="bottom"
                stroke="none"
                text-anchor="middle"
                width="440"
                x="255.55555555555554"
                y="228"
              >
                <tspan
                  dy="0.71em"
                  x="255.55555555555554"
                />
              </text>
            </g>
            <g
              class="recharts-layer recharts-cartesian-axis-tick"
            >
              <text
                class="recharts-text recharts-cartesian-axis-tick-value"
                fill="#666"
                height="30"
                orientation="bottom"
                stroke="none"
                text-anchor="middle"
                width="440"
                x="304.44444444444446"
                y="228"
              >
                <tspan
                  dy="0.71em"
                  x="304.44444444444446"
                />
              </text>
            </g>
            <g
              class="recharts-layer recharts-cartesian-axis-tick"
            >
              <text
                class="recharts-text recharts-cartesian-axis-tick-value"
                fill="#666"
                height="30"
                orientation="bottom"
                stroke="none"
                text-anchor="middle"
                width="440"
                x="353.3333333333333"
                y="228"
              >
                <tspan
                  dy="0.71em"
                  x="353.3333333333333"
                />
              </text>
            </g>
            <g
              class="recharts-layer recharts-cartesian-axis-tick"
            >
              <text
                class="recharts-text recharts-cartesian-axis-tick-value"
                fill="#666"
                height="30"
                orientation="bottom"
                stroke="none"
                text-anchor="middle"
                width="440"
                x="402.2222222222222"
                y="228"
              >
                <tspan
                  dy="0.71em"
                  x="402.2222222222222"
                />
              </text>
            </g>
            <g
              class="recharts-layer recharts-cartesian-axis-tick"
            >
              <text
                class="recharts-text recharts-cartesian-axis-tick-value"
                fill="#666"
                height="30"
                orientation="bottom"
                stroke="none"
                text-anchor="middle"
                width="440"
                x="451.1111111111111"
                y="228"
              >
                <tspan
                  dy="0.71em"
                  x="451.1111111111111"
                />
              </text>
            </g>
            <g
              class="recharts-layer recharts-cartesian-axis-tick"
            >
              <text
                class="recharts-text recharts-cartesian-axis-tick-value"
                fill="#666"
                height="30"
                orientation="bottom"
                stroke="none"
                text-anchor="middle"
                width="440"
                x="500"
                y="228"
              >
                <tspan
                  dy="0.71em"
                  x="500"
                />
              </text>
            </g>
          </g>
        </g>
        <g
          class="recharts-layer recharts-line"
        >
          <path
            class="recharts-curve recharts-line-curve"
            fill="none"
            height="185"
            stroke="#006AFF"
            stroke-dasharray="0px 0px"
            stroke-width="1"
            width="440"
          />
        </g>
        <g
          class="recharts-layer recharts-area"
        >
          <g
            class="recharts-layer"
          >
            <defs>
              <clippath
                id="animationClipPath-recharts-area-6"
              >
                <rect
                  height="2"
                  width="0"
                  x="60"
                  y="0"
                />
              </clippath>
            </defs>
            <g
              class="recharts-layer"
              clip-path="url(#animationClipPath-recharts-area-6)"
            >
              <g
                class="recharts-layer"
              >
                <path
                  class="recharts-curve recharts-area-area"
                  fill="url(#color-value)"
                  fill-opacity="1"
                  height="185"
                  stroke="none"
                  stroke-width="2"
                  width="440"
                />
                <path
                  class="recharts-curve recharts-area-curve"
                  fill="none"
                  fill-opacity="1"
                  height="185"
                  stroke="#006AFF"
                  stroke-width="2"
                  width="440"
                />
              </g>
            </g>
          </g>
        </g>
        <defs>
          <lineargradient
            id="color-value"
            x1="0"
            x2="0"
            y1="0"
            y2="1"
          >
            <stop
              offset="5%"
              stop-color="#006AFF"
              stop-opacity="0.3"
            />
            <stop
              offset="95%"
              stop-color="#006AFF"
              stop-opacity="0"
            />
          </lineargradient>
        </defs>
        <g
          class="recharts-layer recharts-line"
        >
          <path
            class="recharts-curve recharts-line-curve"
            fill="none"
            height="185"
            stroke="#656575"
            stroke-dasharray="0px, 0px"
            stroke-width="1"
            width="440"
          />
        </g>
      </svg>
      <div
        class="recharts-tooltip-wrapper"
        role="dialog"
        style="transform: translate(undefinedpx, undefinedpx); pointer-events: none; visibility: hidden; position: absolute; top: 0px; left: 0px;"
        tabindex="-1"
      >
        <div
          class="recharts-default-tooltip"
          style="margin: 0px; padding: 10px; background-color: rgb(255, 255, 255); border: 1px solid #ccc; white-space: nowrap;"
        >
          <p
            class="recharts-tooltip-label"
            style="margin: 0px;"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
